doc/mainstay.md

```diff
@@ -0,0 +1,127 @@
+# Mainstay integration
+
+Options for attesting state to a mainstay proof-of-publication service. 
+
+Assumptions: 
+
+Mainstay service is available over http interface (or via SOCKS5 Tor proxy). 
+Mainstay service is available and funded with a valid `token_id` for verifiation. 
+Funding (via LN payment) is performed in advance and out of band for subscrption. (i.e. `token_id` is already performed.)
+Mainstay proofs are stored and made available, but that verification against `bitcoind` and staychain occur separately. 
+
+## Config
+
+Node is configured with the mainstay server URL, the slot index and the authentication token:
+
+```
+pub struct MainstayConfig {
+    url: String,
+    position: u64,
+    token: String,
+}
+```
+
+This can be added to `/src/config.rs`
+
+## Commitment function
+
+Impliementation of a commitment function that performs a POST request to the `/commitment/send` mainstay service route, with payload:
+
+```
+payload = {
+  commitment: commitment,
+  position: 0,
+  token: '4c8c006d-4cee-4fef-8e06-bb8112db6314',
+};
+```
+
+`commitment` is a 32 byte value encoded as a 64 character hex string
+
+This can be performed using the `Reqwest` http client library (as in mercury server), e.g. 
+
+```
+use reqwest;
+
+pub struct Request(reqwest::blocking::RequestBuilder);
+
+impl Request {
+    //Construct a request from the give payload and config
+    pub fn from(
+        payload: Option<&Payload>,
+        command: &String,
+        config: &MainstayConfig,
+        signature: Option<String>,
+    ) -> Result<Self> {
+        //Build request
+        let client = reqwest::blocking::Client::new();
+        let url = reqwest::Url::parse(&format!("{}/{}", config.url(), command))?;
+
+        //If there is a payload this is a 'POST' request, otherwise a 'GET' request
+        let req = match payload {
+            Some(p) => {
+                let payload_str = String::from(serde_json::to_string(&p)?);
+                let payload_enc = encode(payload_str);
+                let mut data = HashMap::new();
+                data.insert("X-MAINSTAY-PAYLOAD", &payload_enc);
+                let sig_str = match signature {
+                    Some(s) => s,
+                    None => String::from(""),
+                };
+                data.insert("X-MAINSTAY-SIGNATURE", &sig_str);
+                client
+                    .post(url)
+                    .header(reqwest::header::CONTENT_TYPE, "application/json")
+                    .json(&data)
+            }
+            None => client
+                .get(url)
+                .header(reqwest::header::CONTENT_TYPE, "application/json"),
+        };
+
+        Ok(Self(req))
+    }
+
+    pub fn send(self) -> std::result::Result<reqwest::blocking::Response, reqwest::Error> {
+        self.0.send()
+    }
+}
+```
+
+## Commitment construction
+
+The node will construct commitments from specified *events* () in `src/events.rs`. 
```

Comment:
I think it should be noted the nostr *events* is just a data communication transport and ideally attestation could scope more generic data payload. Note the events in `src/events.rs` are the `civkitd` internal events, even if partially overlapping with Nostr ones.

Reply:
OK - so we need to define precisely what it is that needs to be committed and proven in a dispute. There's no reason this can't be everything that is stored permanently by the node?

Reply:
> There's no reason this can't be everything that is stored permanently by the node?

I don’t get exactly your question, like assuming everything is forever stored by the node? I think you have denial-of-service if proofs can be freely stored or freely queried by clients. Even if you have subscription, there is a need to a data limit.

Reply:
The question is the general one of what data needs to be committed. So all events are saved in the db when `write_new_event_db` is called - is anything else saved in the db apart from events? Strictly, with a proof-of-publication, you should be able to verify everything that has a commitment made from it - so anything you include in a commitment needs to be stored for as long as you want to be able to prove history. So I was meaning that if all events are saved by default indefinitely, then we commit to each one. Or a subset of these events?

To verify, all data that formed the commitment needs to be available. Should we just hash `Event` or `DbEvent` objects? Seems they can be recreated exactly from the values inserted into the DB?

Reply:
> The question is the general one of what data needs to be committed. So all events are saved in the db when write_new_event_db is called - is anything else saved in the db apart from events?

Other elements to be saved in the DB:
- client
- peers
- event subscriptions

> So I was meaning that if all events are saved by default indefinitely, then we commit to each one

Yes I think too, though note the trade-off in term of disk denial-of-service. E.g I ask you to store a proof-of-publication with no time limit and not regular payment. It’s less an issue with what data is saved than client-server interactions.

> To verify, all data that formed the commitment needs to be available. Should we just hash Event or DbEvent objects?

I think we can just have `Event`, db event is a superset or at least you should be able to recreate event from db data elements (at least not now though in the future yes).